std/assembly/util/uri.ts

```diff
@@ -0,0 +1,266 @@
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, /*
+  skip 128 + 1 always set to '1' tail slots */
+]);
+
+// @ts-ignore: decorator
+@lazy export const URI_RESERVED = memory.data<u8>([
+  /*  skip 32 + 3 always set to '0' head slots
+        */ 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1,
+  1, /* skip 191 always set to '0' tail slots */
+]);
+
+export function encode(src: usize, len: usize, table: usize): usize {
+  if (!len) return src;
+
+  var i: usize = 0, offset: usize = 0, outSize = len << 1;
+  var dst = __new(outSize, idof<String>());
+
+  while (i < len) {
+    let org = i;
+    let c: u32, c1: u32;
+    do {
+      c = <u32>load<u16>(src + (i << 1));
+      if (c - 33 < 94) { // 127 - 33
+        if (load<u8>(table + (c - 33))) break;
+      } else break;
+    } while (++i < len);
+
+    if (i > org) {
+      let size = i - org << 1;
+      if (offset + size > outSize) {
+        outSize = offset + size;
+        dst = __renew(dst, outSize);
+      }
+      if (size == 2) {
+        store<u16>(dst + offset, load<u16>(src + (org << 1)));
+      } else if (size == 4) {
+        store<u32>(dst + offset, load<u32>(src + (org << 1)));
+      } else {
+        memory.copy(
+          dst + offset,
+          src + (org << 1),
+          size
+        );
+      }
+      offset += size;
+      if (i >= len) break;
+    }
+
+    if (c >= 0xD800) {
+      if (c >= 0xDC00 && c <= 0xDFFF) {
+        throw new URIError(E_URI_MALFORMED);
+      }
+      if (c <= 0xDBFF) {
+        if (i >= len) {
+          throw new URIError(E_URI_MALFORMED);
+        }
+        c1 = <u32>load<u16>(src + (++i << 1));
+        if (c1 < 0xDC00 || c1 > 0xDFFF) {
+          throw new URIError(E_URI_MALFORMED);
+        }
+        c = (((c & 0x3FF) << 10) | (c1 & 0x3FF)) + 0x10000;
+      }
+    }
+
+    let estSize = offset + (c < 0x80 ? 1 * 6 : 4 * 6);
+    if (estSize > outSize) {
+      outSize = len > 1 ? estSize << 1 : estSize;
+      dst = __renew(dst, outSize);
+    }
+
+    if (c < 0x80) {
+      storeHex(dst, offset, c);
+      offset += 6;
+    } else {
+      if (c < 0x800) {
+        storeHex(dst, offset, (c >> 6) | 0xC0);
+        offset += 6;
+      } else {
+        if (c < 0x10000) {
+          storeHex(dst, offset, (c >> 12) | 0xE0);
+          offset += 6;
+        } else {
+          storeHex(dst, offset, (c >> 18) | 0xF0);
+          offset += 6;
+          storeHex(dst, offset, (c >> 12 & 0x3F) | 0x80);
+          offset += 6;
+        }
+        storeHex(dst, offset, (c >> 6 & 0x3F) | 0x80);
+        offset += 6;
+      }
+      storeHex(dst, offset, (c & 0x3F) | 0x80);
+      offset += 6;
+    }
+    ++i;
+  }
+  if (outSize > offset) {
+    dst = __renew(dst, offset);
+  }
+  return dst;
+}
+
+export function decode(src: usize, len: usize, component: bool): usize {
+  if (!len) return src;
+
+  var i: usize = 0, offset: usize = 0, org: usize, ch: u32 = 0;
```

Comment:
The declaration of `org` can be moved into the `while` I think, be it only to indicate that it is never used uninitialized.